File: src/test/java/com/truveta/opentoken/io/parquet/PersonAttributesParquetWriterTest.java
/**
 * Copyright (c) Truveta. All rights reserved.
 */
package com.truveta.opentoken.io.parquet;

import static org.junit.jupiter.api.Assertions.*;
import java.io.File;
import java.nio.file.Files;
import java.util.LinkedHashMap;
import java.util.Map;
import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

class PersonAttributesParquetWriterTest {

    private File tempFile;
    private String tempFilePath;
    private PersonAttributesParquetWriter writer;

    @BeforeEach
    void setUp() throws Exception {
        tempFile = Files.createTempFile("test_data", ".parquet").toFile();
        tempFilePath = tempFile.getAbsolutePath();
        writer = new PersonAttributesParquetWriter(tempFilePath);
    }

    @AfterEach
    void tearDown() throws Exception {
        if (writer != null) {
            writer.close();
        }
        if (tempFile.exists()) {
            tempFile.delete();
        }
    }

    @Test
    void testWriteSingleRecord() throws Exception {
        Map<String, String> data = new LinkedHashMap<>();
        data.put("RecordId", "123");
        data.put("Name", "John Doe");
        data.put("SocialSecurityNumber", "123-45-6789");

        writer.writeAttributes(data);
        writer.close();

        try (PersonAttributesParquetReader reader = new PersonAttributesParquetReader(tempFilePath)) {
            boolean hasNext = reader.hasNext();
            assertTrue(hasNext);

            Map<String, String> record = reader.next();
            assertNotNull(record);
            assertEquals("123", record.get("RecordId"));
            assertEquals("123-45-6789", record.get("SocialSecurityNumber"));
            assertEquals("John Doe", record.get("Name"));
        }
    }

    @Test
    void testWriteMultipleRecords() throws Exception {
        Map<String, String> data1 = new LinkedHashMap<>();
        data1.put("RecordId", "123");
        data1.put("Name", "John Doe");
        data1.put("SocialSecurityNumber", "123-45-6789");

        Map<String, String> data2 = new LinkedHashMap<>();
        data2.put("RecordId", "456");
        data2.put("Name", "Jane Smith");
        data2.put("SocialSecurityNumber", "987-65-4321");

        writer.writeAttributes(data1);
        writer.writeAttributes(data2);
        writer.close();

        try (PersonAttributesParquetReader reader = new PersonAttributesParquetReader(tempFilePath)) {
            boolean hasNext = reader.hasNext();
            assertTrue(hasNext);

            Map<String, String> record = reader.next();
            assertNotNull(record);
            assertEquals("123", record.get("RecordId"));
            assertEquals("123-45-6789", record.get("SocialSecurityNumber"));
            assertEquals("John Doe", record.get("Name"));

            hasNext = reader.hasNext();
            assertTrue(hasNext);

            record = reader.next();
            assertNotNull(record);
            assertEquals("456", record.get("RecordId"));
            assertEquals("987-65-4321", record.get("SocialSecurityNumber"));
            assertEquals("Jane Smith", record.get("Name"));
        }
    }
}

File: lib/python/openlinktoken-cli/src/test/openlinktoken_cli/io/parquet/person_attributes_parquet_writer_test.py
# SPDX-License-Identifier: MIT

import os
import tempfile
from pathlib import Path

from openlinktoken_cli.io.parquet.person_attributes_parquet_reader import PersonAttributesParquetReader
from openlinktoken_cli.io.parquet.person_attributes_parquet_writer import PersonAttributesParquetWriter


class TestPersonAttributesParquetWriter:
    """Test cases for PersonAttributesParquetWriter."""

    def setup_method(self):
        """Set up test fixtures before each test method."""
        self.temp_file = tempfile.NamedTemporaryFile(suffix=".parquet", delete=False)
        self.temp_file_path = self.temp_file.name
        self.temp_file.close()
        self.writer = PersonAttributesParquetWriter(self.temp_file_path)

    def teardown_method(self):
        """Clean up after each test method."""
        if self.writer:
            self.writer.close()
        if os.path.exists(self.temp_file_path):
            os.unlink(self.temp_file_path)

    def test_write_single_record(self):
        """Test writing a single record to Parquet."""
        data = {"RecordId": "123", "FirstName": "John", "SocialSecurityNumber": "123-45-6789"}

        self.writer.write_attributes(data)
        self.writer.close()

        with PersonAttributesParquetReader(self.temp_file_path) as reader:
            record = next(reader)
            assert record is not None
            assert record["RecordId"] == "123"
            assert record["SocialSecurityNumber"] == "123-45-6789"
            assert record["FirstName"] == "John"

    def test_write_multiple_records(self):
        """Test writing multiple records to Parquet."""
        data1 = {"RecordId": "123", "FirstName": "John", "SocialSecurityNumber": "123-45-6789"}
        data2 = {"RecordId": "456", "FirstName": "Jane", "SocialSecurityNumber": "987-65-4321"}

        self.writer.write_attributes(data1)
        self.writer.write_attributes(data2)
        self.writer.close()

        with PersonAttributesParquetReader(self.temp_file_path) as reader:
            # Test first record
            record = next(reader)
            assert record is not None
            assert record["RecordId"] == "123"
            assert record["SocialSecurityNumber"] == "123-45-6789"
            assert record["FirstName"] == "John"

            # Test second record
            record = next(reader)
            assert record is not None
            assert record["RecordId"] == "456"
            assert record["SocialSecurityNumber"] == "987-65-4321"
            assert record["FirstName"] == "Jane"

    def test_write_basename_output_path_in_current_directory(self, tmp_path, monkeypatch):
        """Test that a basename-only output path writes to the current directory."""
        monkeypatch.chdir(tmp_path)
        writer = PersonAttributesParquetWriter("output.parquet")

        try:
            writer.write_attributes({"RecordId": "123", "FirstName": "John"})
        finally:
            writer.close()

        output_path = tmp_path / "output.parquet"
        assert output_path.exists()
